token-manager/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+        self.only_service();
+
+        let token_identifier = self.token_identifier().get();
+
+        self.add_flow_in_raw(&amount);
+
+        let implementation_type = self.implementation_type().get();
+        match implementation_type {
+            TokenManagerType::MintBurn | TokenManagerType::MintBurnFrom => {
+                self.give_token_mint_burn(&token_identifier, destination_address, &amount);
+            }
+            // nothing to do for lock/unlock, tokens remain in contract
+            TokenManagerType::LockUnlock | TokenManagerType::LockUnlockFee => {
+                self.give_token_lock_unlock(&token_identifier, destination_address, &amount);
+            }
+        }
+
+        (token_identifier, amount).into()
+    }
+
+    #[payable("*")]
+    #[endpoint(takeToken)]
+    fn take_token(&self) -> BigUint {
+        self.only_service();
+
+        let (token_identifier, amount) = self.require_correct_token();
+
+        self.add_flow_out_raw(&amount);
+
+        let implementation_type = self.implementation_type().get();
+        match implementation_type {
+            TokenManagerType::MintBurn | TokenManagerType::MintBurnFrom => {
+                self.take_token_mint_burn(token_identifier, &amount);
+            }
+            // nothing to do for lock/unlock, tokens remain in contract
+            TokenManagerType::LockUnlock | TokenManagerType::LockUnlockFee => {}
+        }
+
+        amount
+    }
+
+    /// Mint/burn type only functions
+
+    // Somewhat equivalent to Axelar InterchainToken init method
+    #[payable("EGLD")]
+    #[endpoint(deployInterchainToken)]
+    fn deploy_interchain_token(
+        &self,
+        minter: Option<ManagedAddress>,
+        name: ManagedBuffer,
+        symbol: ManagedBuffer,
+        decimals: u8,
+    ) {
+        require!(
+            self.implementation_type().get() == TokenManagerType::MintBurn,
+            "Not mint burn token manager"
+        );
+
+        require!(
+            self.token_identifier().is_empty(),
+            "Token address already exists"
+        );
+
+        let caller = self.blockchain().get_caller();
+        let interchain_token_service = self.interchain_token_service().get();
+
+        // Also allow minter to call this (if set) in case issue esdt fails
+        require!(
+            caller == interchain_token_service || self.is_minter(&caller),
+            "Not service or minter"
+        );
+
+        require!(!name.is_empty(), "Token name empty");
+        require!(!symbol.is_empty(), "Token symbol empty");
+
+        /*
+         * Set the token service as a minter to allow it to mint and burn tokens.
+         * Also add the provided address as a minter. If zero address was provided,
+         * add it as a minter to allow anyone to easily check that no custom minter was set.
+         */
+        self.add_minter(interchain_token_service);
+        if minter.is_some() {
```

Comment:
There might be an issue here with `minter` because there is a module name `minter` and also the variable. Not sure if this works.

Reply:
Changed the name of the `minter` module to `mintership` to better reflect the functionality.